### PR TITLE
feat: add erc20 proxy with fee payment processor

### PR DESCRIPTION
**Pull request by rittme**
## Description of the changes
Add payment-processor support to ERC20 with fees proxy contract

### CODE REVIEW

File: packages/payment-processor/src/payment/utils.ts [hunk 1]
```diff
@@ -111,7 +116,10 @@ export function validateRequest(
     !extension ||
     !extension.values.salt ||
     !extension.values.paymentAddress ||
-    (paymentNetworkId === ERC20_PROXY_CONTRACT && !request.currencyInfo.value)
+    (paymentNetworkId === ERC20_PROXY_CONTRACT && !request.currencyInfo.value) ||
+    (paymentNetworkId === ERC20_FEE_PROXY_CONTRACT && !request.currencyInfo.value) ||
+    (paymentNetworkId === ERC20_FEE_PROXY_CONTRACT &&
+      !!extension.values.feeAddress !== !!extension.values.feeAmount)
```
Comment (benjlevesque) [inline packages/payment-processor/src/payment/utils.ts:122]: aren't those 2 mandatory for an `ERC20_FEE_PROXY_CONTRACT`?

File: packages/payment-processor/test/payment/erc20-fee-proxy.test.ts [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    transactionManagerMeta: {},
+  },
+  pending: null,
+  requestId: 'abcd',
+  state: RequestLogicTypes.STATE.CREATED,
+  timestamp: 0,
+  version: '1.0',
+};
+
+describe.only('erc20-fee-proxy', () => {
```
Comment (benjlevesque) [inline packages/payment-processor/test/payment/erc20-fee-proxy.test.ts:87]:
```suggestion
describe('erc20-fee-proxy', () => {
```

File: packages/payment-processor/src/payment/erc20-fee-proxy.ts [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+  });
+  return tx;
+}
+
+/**
+ * Encodes the approval call, can be used with a Multisig contract.
+ * @param request the request to pay
+ * @param signerOrProvider the Web3 provider, or signer. Defaults to window.ethereum.
+ */
+export function encodeApproveErc20(
```
Comment (benjlevesque) [inline packages/payment-processor/src/payment/erc20-fee-proxy.ts:137]: maybe this method can be reused from the other PN? 
only the `validateRequest` changes, I believe, and as it's checked on the payment, maybe it's enough?

File: packages/payment-processor/src/payment/erc20-fee-proxy.ts [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+  return allowance.gt(bigNumberify(request.expectedAmount).add(feeAmount || 0));
+}
+
+/**
+ * Processes the approval transaction of the targeted ERC20.
+ * @param request request to pay
+ * @param provider the web3 provider. Defaults to Etherscan.
+ * @param overrides optionally, override default transaction values, like gas.
+ */
+export async function approveErc20(
```
Comment (benjlevesque) [inline packages/payment-processor/src/payment/erc20-fee-proxy.ts:115]: same as below

File: packages/payment-processor/src/payment/erc20-fee-proxy.ts [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+ * @param request
+ * @param signerOrProvider the Web3 provider, or signer. Defaults to window.ethereum.
+ * @param amount optionally, the amount to pay. Defaults to remaining amount of the request.
+ * @param overrides optionally, override default transaction values, like gas.
+ */
+export async function payErc20FeeProxyRequest(
+  request: ClientTypes.IRequestData,
+  signerOrProvider: Web3Provider | Signer = getProvider(),
+  amount?: BigNumberish,
+  feeAmount?: BigNumberish,
```
Comment (benjlevesque) [inline packages/payment-processor/src/payment/erc20-fee-proxy.ts:32]: misses documentation

File: packages/payment-processor/src/payment/erc20-fee-proxy.ts [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+ * Encodes the call to pay a request through the ERC20 fee proxy contract, can be used with a Multisig contract.
+ * @param request request to pay
+ * @param signerOrProvider the Web3 provider, or signer. Defaults to window.ethereum.
+ * @param amount optionally, the amount to pay. Defaults to remaining amount of the request.
+ */
+export function encodePayErc20FeeRequest(
+  request: ClientTypes.IRequestData,
+  signerOrProvider: Web3Provider | Signer = getProvider(),
+  amount?: BigNumberish,
+  feeAmountOverride?: BigNumberish,
```
Comment (benjlevesque) [inline packages/payment-processor/src/payment/erc20-fee-proxy.ts:58]: misses documentation

File: packages/payment-processor/src/payment/utils.ts (hunk 1 above)
Comment (rittme) [reply packages/payment-processor/src/payment/utils.ts:122]: no, it's mandatory to have both, or neither. If neither, it just behaves like a normal erc20 proxy.

File: packages/payment-processor/src/payment/erc20-fee-proxy.ts (hunk 3 above)
Comment (rittme) [reply packages/payment-processor/src/payment/erc20-fee-proxy.ts:137]: It's also the address. I'll make it more generic on the other PN and reuse.

**Pull request merged**